File: app/src/main/kotlin/tech/mmarca/openvitals/features/nutrition/NutritionFormatting.kt
package tech.mmarca.openvitals.features.nutrition

import androidx.compose.ui.graphics.Color
import tech.mmarca.openvitals.R
import tech.mmarca.openvitals.core.presentation.DisplayValue
import tech.mmarca.openvitals.core.presentation.UnitFormatter
import tech.mmarca.openvitals.domain.model.NutritionNutrient
import tech.mmarca.openvitals.domain.model.NutritionNutrientGroup
import tech.mmarca.openvitals.domain.model.NutritionNutrientUnit
import tech.mmarca.openvitals.ui.components.PeriodChartValue
import tech.mmarca.openvitals.ui.theme.NutritionColor
import kotlin.math.roundToInt

private val proteinMetricColor = Color(0xFF7E57C2)
private val carbsMetricColor = Color(0xFF26A69A)
private val fatMetricColor = Color(0xFFFFB300)

internal data class NutritionSeriesUiModel(
    val nutrient: NutritionNutrient,
    val titleRes: Int,
    val total: DisplayValue,
    val totalValue: Double,
    val values: List<PeriodChartValue>,
    val color: Color,
    val valueDisplayFormatter: (Double) -> DisplayValue,
) {
    val hasTrackedValues: Boolean = values.any { it.value > 0.0 }
}

internal fun NutritionNutrientSeries.toUiModel(unitFormatter: UnitFormatter): NutritionSeriesUiModel =
    NutritionSeriesUiModel(
        nutrient = nutrient,
        titleRes = nutrient.titleRes(),
        total = nutrient.displayValue(totalValue, unitFormatter),
        totalValue = totalValue,
        values = values.map { PeriodChartValue(date = it.date, value = it.value) },
        color = nutrient.color(),
        valueDisplayFormatter = { nutrient.displayValue(it, unitFormatter) },
    )

internal fun NutritionMetricDisplay.toUiModel(unitFormatter: UnitFormatter): NutritionSeriesUiModel =
    NutritionSeriesUiModel(
        nutrient = nutrient,
        titleRes = nutrient.titleRes(),
        total = nutrient.displayValue(totalValue, unitFormatter),
        totalValue = totalValue,
        values = values.map { PeriodChartValue(date = it.date, value = it.value) },
        color = nutrient.color(),
        valueDisplayFormatter = { nutrient.displayValue(it, unitFormatter) },
    )

internal fun NutritionNutrient.displayValue(
    value: Double,
    unitFormatter: UnitFormatter,
): DisplayValue =
    when (unit) {
        NutritionNutrientUnit.ENERGY_KCAL -> unitFormatter.energy(value)
        NutritionNutrientUnit.MASS_GRAMS -> DisplayValue(unitFormatter.count(value.roundToInt()), "g")
        NutritionNutrientUnit.MASS_ADAPTIVE -> adaptiveMassDisplay(value, unitFormatter)
    }

private fun adaptiveMassDisplay(
    grams: Double,
    unitFormatter: UnitFormatter,
): DisplayValue {
    val milligrams = grams * 1_000.0
    val micrograms = grams * 1_000_000.0
    return when {
        grams >= 1.0 -> DisplayValue(unitFormatter.decimal(grams, if (grams < 10.0) 1 else 0), "g")
        milligrams >= 1.0 -> DisplayValue(unitFormatter.decimal(milligrams, if (milligrams < 10.0) 1 else 0), "mg")
        else -> DisplayValue(unitFormatter.decimal(micrograms, if (micrograms < 10.0) 1 else 0), "mcg")
    }
}

internal fun NutritionNutrient.color(): Color =
    when (group) {
        NutritionNutrientGroup.OVERVIEW -> when (this) {
            NutritionNutrient.ENERGY -> NutritionColor
            NutritionNutrient.PROTEIN -> proteinMetricColor
            NutritionNutrient.TOTAL_CARBOHYDRATE -> carbsMetricColor
            NutritionNutrient.TOTAL_FAT -> fatMetricColor
            else -> NutritionColor
        }
        NutritionNutrientGroup.CARBOHYDRATES -> carbsMetricColor
        NutritionNutrientGroup.FATS -> fatMetricColor
        NutritionNutrientGroup.VITAMINS -> Color(0xFF5E7CE2)
        NutritionNutrientGroup.MINERALS -> Color(0xFF8D6E63)
        NutritionNutrientGroup.OTHER -> Color(0xFF00897B)
    }

internal fun NutritionNutrientGroup.titleRes(): Int =
    when (this) {
        NutritionNutrientGroup.OVERVIEW -> R.string.screen_nutrition
        NutritionNutrientGroup.CARBOHYDRATES -> R.string.section_carbohydrates
        NutritionNutrientGroup.FATS -> R.string.section_fats
        NutritionNutrientGroup.VITAMINS -> R.string.section_vitamins
        NutritionNutrientGroup.MINERALS -> R.string.section_minerals
        NutritionNutrientGroup.OTHER -> R.string.section_other_nutrients
    }

internal fun NutritionNutrient.titleRes(): Int =
    when (this) {
        NutritionNutrient.ENERGY -> R.string.metric_calories_in
        NutritionNutrient.PROTEIN -> R.string.metric_protein
        NutritionNutrient.TOTAL_CARBOHYDRATE -> R.string.metric_carbs
        NutritionNutrient.TOTAL_FAT -> R.string.metric_fat
        NutritionNutrient.DIETARY_FIBER -> R.string.metric_dietary_fiber
        NutritionNutrient.SUGAR -> R.string.metric_sugar
        NutritionNutrient.ENERGY_FROM_FAT -> R.string.metric_energy_from_fat
        NutritionNutrient.MONOUNSATURATED_FAT -> R.string.metric_monounsaturated_fat
        NutritionNutrient.POLYUNSATURATED_FAT -> R.string.metric_polyunsaturated_fat
        NutritionNutrient.SATURATED_FAT -> R.string.metric_saturated_fat
        NutritionNutrient.TRANS_FAT -> R.string.metric_trans_fat
        NutritionNutrient.UNSATURATED_FAT -> R.string.metric_unsaturated_fat
        NutritionNutrient.CHOLESTEROL -> R.string.metric_cholesterol
        NutritionNutrient.BIOTIN -> R.string.metric_biotin
        NutritionNutrient.FOLATE -> R.string.metric_folate
        NutritionNutrient.FOLIC_ACID -> R.string.metric_folic_acid
        NutritionNutrient.NIACIN -> R.string.metric_niacin
        NutritionNutrient.PANTOTHENIC_ACID -> R.string.metric_pantothenic_acid
        NutritionNutrient.RIBOFLAVIN -> R.string.metric_riboflavin
        NutritionNutrient.THIAMIN -> R.string.metric_thiamin
        NutritionNutrient.VITAMIN_A -> R.string.metric_vitamin_a
        NutritionNutrient.VITAMIN_B12 -> R.string.metric_vitamin_b12
        NutritionNutrient.VITAMIN_B6 -> R.string.metric_vitamin_b6
        NutritionNutrient.VITAMIN_C -> R.string.metric_vitamin_c
        NutritionNutrient.VITAMIN_D -> R.string.metric_vitamin_d
        NutritionNutrient.VITAMIN_E -> R.string.metric_vitamin_e
        NutritionNutrient.VITAMIN_K -> R.string.metric_vitamin_k
        NutritionNutrient.CALCIUM -> R.string.metric_calcium
        NutritionNutrient.CHLORIDE -> R.string.metric_chloride
        NutritionNutrient.CHROMIUM -> R.string.metric_chromium
        NutritionNutrient.COPPER -> R.string.metric_copper
        NutritionNutrient.IODINE -> R.string.metric_iodine
        NutritionNutrient.IRON -> R.string.metric_iron
        NutritionNutrient.MAGNESIUM -> R.string.metric_magnesium
        NutritionNutrient.MANGANESE -> R.string.metric_manganese
        NutritionNutrient.MOLYBDENUM -> R.string.metric_molybdenum
        NutritionNutrient.PHOSPHORUS -> R.string.metric_phosphorus
        NutritionNutrient.POTASSIUM -> R.string.metric_potassium
        NutritionNutrient.SELENIUM -> R.string.metric_selenium
        NutritionNutrient.SODIUM -> R.string.metric_sodium
        NutritionNutrient.ZINC -> R.string.metric_zinc
        NutritionNutrient.CAFFEINE -> R.string.metric_caffeine
    }

File: lib/features/nutrition/presentation/nutrition_formatting.dart
import 'package:flutter/material.dart';

import '../../../core/presentation/display_value.dart';
import '../../../core/presentation/unit_formatter.dart';
import '../../../domain/model/nutrition_models.dart';
import '../../../l10n/app_localizations.dart';
import '../../../ui/theme/app_colors.dart';

/// Per-nutrient presentation helpers, ported from the Kotlin
/// `NutritionFormatting`.

const Color _proteinColor = Color(0xFF7E57C2);
const Color _carbsColor = Color(0xFF26A69A);
const Color _fatColor = Color(0xFFFFB300);
const Color _vitaminColor = Color(0xFF5E7CE2);
const Color _mineralColor = Color(0xFF8D6E63);
const Color _otherColor = Color(0xFF00897B);

/// Formats [value] for [nutrient] using its unit (energy / mass grams /
/// adaptive mass), mirroring the Kotlin `NutritionNutrient.displayValue`.
DisplayValue nutrientDisplayValue(
  NutritionNutrient nutrient,
  double value,
  UnitFormatter formatter,
) {
  switch (nutrient.unit) {
    case NutritionNutrientUnit.energyKcal:
      return formatter.energy(value);
    case NutritionNutrientUnit.massGrams:
      return DisplayValue(formatter.count(value.round()), 'g');
    case NutritionNutrientUnit.massAdaptive:
      return _adaptiveMassDisplay(value, formatter);
  }
}

DisplayValue _adaptiveMassDisplay(double grams, UnitFormatter formatter) {
  final milligrams = grams * 1000.0;
  final micrograms = grams * 1000000.0;
  if (grams >= 1.0) {
    return DisplayValue(formatter.decimal(grams, grams < 10.0 ? 1 : 0), 'g');
  }
  if (milligrams >= 1.0) {
    return DisplayValue(
      formatter.decimal(milligrams, milligrams < 10.0 ? 1 : 0),
      'mg',
    );
  }
  return DisplayValue(
    formatter.decimal(micrograms, micrograms < 10.0 ? 1 : 0),
    'mcg',
  );
}

/// The accent colour for [nutrient], mirroring the Kotlin
/// `NutritionNutrient.color`.
Color nutrientColor(NutritionNutrient nutrient) {
  switch (nutrient.group) {
    case NutritionNutrientGroup.overview:
      switch (nutrient) {
        case NutritionNutrient.energy:
          return AppColors.nutrition;
        case NutritionNutrient.protein:
          return _proteinColor;
        case NutritionNutrient.totalCarbohydrate:
          return _carbsColor;
        case NutritionNutrient.totalFat:
          return _fatColor;
        default:
          return AppColors.nutrition;
      }
    case NutritionNutrientGroup.carbohydrates:
      return _carbsColor;
    case NutritionNutrientGroup.fats:
      return _fatColor;
    case NutritionNutrientGroup.vitamins:
      return _vitaminColor;
    case NutritionNutrientGroup.minerals:
      return _mineralColor;
    case NutritionNutrientGroup.other:
      return _otherColor;
  }
}

/// The display title for [group], mirroring the Kotlin
/// `NutritionNutrientGroup.titleRes`.
String nutrientGroupTitle(NutritionNutrientGroup group) {
  switch (group) {
    case NutritionNutrientGroup.overview:
      return 'Nutrition';
    case NutritionNutrientGroup.carbohydrates:
      return 'Carbohydrates';
    case NutritionNutrientGroup.fats:
      return 'Fats';
    case NutritionNutrientGroup.vitamins:
      return 'Vitamins';
    case NutritionNutrientGroup.minerals:
      return 'Minerals';
    case NutritionNutrientGroup.other:
      return 'Other nutrients';
  }
}

/// The localized display title for [nutrient]. Port of the Kotlin
/// `NutritionNutrient.titleRes()` — the same `metric_*` strings, so a nutrient
/// reads identically wherever it appears.
String nutrientTitle(NutritionNutrient nutrient, AppLocalizations l10n) =>
    switch (nutrient) {
      NutritionNutrient.energy => l10n.metricCaloriesIn,
      NutritionNutrient.protein => l10n.metricProtein,
      NutritionNutrient.totalCarbohydrate => l10n.metricCarbs,
      NutritionNutrient.totalFat => l10n.metricFat,
      NutritionNutrient.dietaryFiber => l10n.metricDietaryFiber,
      NutritionNutrient.sugar => l10n.metricSugar,
      NutritionNutrient.energyFromFat => l10n.metricEnergyFromFat,
      NutritionNutrient.monounsaturatedFat => l10n.metricMonounsaturatedFat,
      NutritionNutrient.polyunsaturatedFat => l10n.metricPolyunsaturatedFat,
      NutritionNutrient.saturatedFat => l10n.metricSaturatedFat,
      NutritionNutrient.transFat => l10n.metricTransFat,
      NutritionNutrient.unsaturatedFat => l10n.metricUnsaturatedFat,
      NutritionNutrient.cholesterol => l10n.metricCholesterol,
      NutritionNutrient.biotin => l10n.metricBiotin,
      NutritionNutrient.folate => l10n.metricFolate,
      NutritionNutrient.folicAcid => l10n.metricFolicAcid,
      NutritionNutrient.niacin => l10n.metricNiacin,
      NutritionNutrient.pantothenicAcid => l10n.metricPantothenicAcid,
      NutritionNutrient.riboflavin => l10n.metricRiboflavin,
      NutritionNutrient.thiamin => l10n.metricThiamin,
      NutritionNutrient.vitaminA => l10n.metricVitaminA,
      NutritionNutrient.vitaminB12 => l10n.metricVitaminB12,
      NutritionNutrient.vitaminB6 => l10n.metricVitaminB6,
      NutritionNutrient.vitaminC => l10n.metricVitaminC,
      NutritionNutrient.vitaminD => l10n.metricVitaminD,
      NutritionNutrient.vitaminE => l10n.metricVitaminE,
      NutritionNutrient.vitaminK => l10n.metricVitaminK,
      NutritionNutrient.calcium => l10n.metricCalcium,
      NutritionNutrient.chloride => l10n.metricChloride,
      NutritionNutrient.chromium => l10n.metricChromium,
      NutritionNutrient.copper => l10n.metricCopper,
      NutritionNutrient.iodine => l10n.metricIodine,
      NutritionNutrient.iron => l10n.metricIron,
      NutritionNutrient.magnesium => l10n.metricMagnesium,
      NutritionNutrient.manganese => l10n.metricManganese,
      NutritionNutrient.molybdenum => l10n.metricMolybdenum,
      NutritionNutrient.phosphorus => l10n.metricPhosphorus,
      NutritionNutrient.potassium => l10n.metricPotassium,
      NutritionNutrient.selenium => l10n.metricSelenium,
      NutritionNutrient.sodium => l10n.metricSodium,
      NutritionNutrient.zinc => l10n.metricZinc,
      NutritionNutrient.caffeine => l10n.metricCaffeine,
    };
